src/otx/v2/adapters/torch/mmengine/dataset.py

```diff
@@ -0,0 +1,310 @@
+"""OTX adapters.torch.mmengine.Dataset API."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from functools import partial
+from typing import TYPE_CHECKING, Iterable
+
+from mmengine.dataset import default_collate, worker_init_fn
+from mmengine.dist import get_dist_info
+
+from otx.v2.adapters.torch.dataset import TorchBaseDataset
+from otx.v2.adapters.torch.mmengine.modules.utils.config_utils import CustomConfig as Config
+from otx.v2.adapters.torch.mmengine.registry import MMEngineRegistry
+from otx.v2.adapters.torch.modules.dataloaders import ComposedDL
+from otx.v2.api.entities.task_type import TaskType, TrainType
+from otx.v2.api.utils.decorators import add_subset_dataloader
+from otx.v2.api.utils.type_utils import str_to_subset_type
+
+if TYPE_CHECKING:
+    from mmengine.registry import Registry
+    from torch.utils.data import DataLoader as TorchDataLoader
+    from torch.utils.data import Dataset as TorchDataset
+    from torch.utils.data import Sampler
+
+SUBSET_LIST = ["train", "val", "test"]
+
+
+@add_subset_dataloader(SUBSET_LIST)
+class MMXDataset(TorchBaseDataset):
```

Comment:
I don't know why there are many different `Dataset` for each framework. What's the reason? Is it for the input type transformation for the model developed by each framework?

Reply:
In the case of MMXDataset, yes, it's meant to build by converting inputs, and we try to keep the common code together as much as possible. And each of mmpretrain, mmdet, mmseg, etc. have their own overrides that are needed because they all use modules that turn a DatasetEntity into a Dataset that can be used in mmlab. I'm hoping that eliminating DatasetEntity will reduce the number of unnecessary classes.

Reply:
Yes, I think that's a good suggestion but there are a few things to check. Currently, we utilize the registry that each mmlab library uses to build the dataset, so this could be a conflict from there, but It would be great to discuss and think about this a bit more.